packages/notebook/src/history.ts

```diff
@@ -0,0 +1,365 @@
+   */
+  updateEditor(activeCell: Cell, update: string): void {
+    if (activeCell) {
+      const model = activeCell.editor?.model;
+      const source = model?.sharedModel.getSource();
+      if (this.isDisposed || !update) {
+        return;
+      }
+      if (source === update) {
+        return;
+      }
+      this._setByHistory = true;
+      model?.sharedModel.setSource(update);
+      let columnPos = 0;
+      columnPos = update.indexOf('\n');
+      if (columnPos < 0) {
+        columnPos = update.length;
+      }
+      activeCell.editor?.setCursorPosition({ line: 0, column: columnPos });
+    }
+  }
+
+  /**
+   * Reset the history navigation state, i.e., start a new history session.
+   */
+  reset(): void {
+    this._hasSession = false;
+    this._placeholder = '';
+  }
+
+  /**
+   * Populate the history collection on history reply from a kernel.
+   *
+   * @param value The kernel message history reply.
+   *
+   * #### Notes
+   * History entries have the shape:
+   * [session: number, line: number, input: string]
+   * Contiguous duplicates are stripped out of the API response.
+   */
+  protected onHistory(
+    value: KernelMessage.IHistoryReplyMsg,
+    cell?: Cell
+  ): void {
+    this._history.length = 0;
+    let last = ['', '', ''];
+    let current = ['', '', ''];
+    let kernelSession = '';
+    if (value.content.status === 'ok') {
+      for (let i = 0; i < value.content.history.length; i++) {
+        current = value.content.history[i] as string[];
+        if (current !== last) {
+          kernelSession = (value.content.history[i] as string[])[0];
+          this._history.push((last = current));
+        }
+      }
+      // set the kernel session for filtering
+      if (!this.kernelSession) {
+        if (current[2] == cell?.model.sharedModel.getSource()) {
+          this._kernelSession = kernelSession;
+        }
+      }
+    }
+  }
+
+  /**
+   * Handle a text change signal from the editor.
+   */
+  protected onTextChange(): void {
+    if (this._setByHistory) {
+      this._setByHistory = false;
+      return;
+    }
+    this.reset();
+  }
+
+  /**
+   * Handle the current kernel changing.
+   */
+  private async _handleKernel(): Promise<void> {
+    this._kernel = this.sessionContext.session?.kernel;
+    if (!this._kernel) {
+      this._history.length = 0;
+      return;
+    }
+    await this._retrieveHistory().catch();
+    return;
+  }
+
+  /**
+   * handle a cell execution
+   */
+  private async _onExecuted(
+    _: unknown,
+    args: { notebook: StaticNotebook; cell: Cell }
+  ): Promise<void> {
+    const cell = args['cell'];
+    await this._retrieveHistory(cell).catch();
+  }
+
+  /**
+   * retrieve the history from the kernel
+   *
+   * @param cell - The string to use when filtering the data.
+   */
+  private async _retrieveHistory(cell?: Cell): Promise<void> {
+    return await this._kernel
+      ?.requestHistory(Private.initialRequest)
```

Comment:
Any suggestion on how this should be handled? It seems easy enough to catch this, but I'm wondering how it should be handled after the catch. Would a notification.warning with autoClose be sufficient?

Reply:
I would say that this is even fine to have as `console.warn` - this is a bit pathological case of kernel dying; it is just a bad practice to have uncaught exceptions because these sometimes cause trouble to users with extensions using catch-all handlers (and make user bug reports noisy).